server/services/odysee.ts

```diff
@@ -0,0 +1,794 @@
+		lbry += `/${slug}${streamId ? `#${streamId}` : ""}`;
+	}
+	return lbry;
+}
+
+// Normalize thumbnail field(s) into a list of URLs
+function hasProp<T extends PropertyKey>(o: unknown, k: T): o is Record<T, unknown> {
+	return !!o && typeof o === "object" && k in o;
+}
+
+function getSdHashFromValue(val: unknown): string | undefined {
+	if (hasProp(val, "source") && hasProp(val.source, "sd_hash")) {
+		const sd = val.source.sd_hash;
+		return typeof sd === "string" ? sd : undefined;
+	}
+	return undefined;
+}
+
+function extractThumbnails(thumb: unknown): string[] {
+	const out: string[] = [];
+	if (typeof thumb === "string") {
+		out.push(thumb);
+	} else if (hasProp(thumb, "url")) {
+		const u = thumb.url;
+		if (typeof u === "string" && u) {
+			out.push(u);
+		}
+	}
+	return out;
+}
+
+// Collect thumbnails with fallbacks: claim -> channel -> lbry.com
+function collectThumbnails(resolved: ResolveMap[string] | undefined, got: LbryGetResult): string[] {
+	const v = got.value ?? resolved?.value ?? {};
+	let thumbnails = extractThumbnails(v.thumbnail);
+	if (thumbnails.length === 0 && resolved?.value?.thumbnail) {
+		thumbnails = extractThumbnails(resolved.value.thumbnail);
+	}
+	if (thumbnails.length === 0) {
+		const chVal = got.signing_channel?.value ?? resolved?.signing_channel?.value;
+		if (chVal?.thumbnail) {
+			thumbnails = extractThumbnails(chVal.thumbnail);
+		}
+		if (thumbnails.length === 0 && chVal?.cover) {
+			thumbnails = extractThumbnails(chVal.cover);
+		}
+	}
+	if (thumbnails.length === 0) {
+		const cid = resolved?.claim_id || got.claim_id;
+		if (cid) {
+			thumbnails.push(`https://thumbnails.lbry.com/${cid}`);
+		}
+	}
+	return thumbnails;
+}
+
+function isOdyCdnHost(u: string): boolean {
+	try {
+		return /(\.|^)odycdn\.com$/i.test(new URL(u).hostname);
+	} catch {
+		return false;
+	}
+}
+
+// Detect HLS by MIME or URL pattern
+function isHlsMimeOrUrl(mime?: string | null, url?: string): boolean {
+	const m = (mime || "").toLowerCase();
+	if (
+		m === "application/x-mpegurl" ||
+		m === "application/vnd.apple.mpegurl" ||
+		m === "application/mpegurl" ||
+		m === "audio/mpegurl"
+	) {
+		return true;
+	}
+	if (url) {
+		const clean = url.split("#")[0].split("?")[0];
+		const filename = clean.split("/").pop() ?? "";
+		const ext = (filename.split(".").pop() ?? "").toLowerCase();
+		if (ext && getMimeType(ext) === "application/x-mpegURL") {
+			return true;
+		}
+	}
+	return false;
+}
+
+// Canonicalize MIME; map HLS to application/x-mpegURL; default to MP4
+function normalizeMime(mimeFromServer?: string | null, url?: string): string {
+	if (isHlsMimeOrUrl(mimeFromServer, url)) {
+		return "application/x-mpegURL";
+	}
+	return mimeFromServer?.startsWith("video/") ? mimeFromServer : "video/mp4";
+}
+
+// Make absolute URL using base if needed
+function resolveRelativeUrl(base: string, maybeRelative: string): string {
+	try {
+		return new URL(maybeRelative, base).toString();
```

Comment:
Seems a bit silly to keep parsing the same strings as URLs over and over... Consider parsing it as a URL once and do operations on the URL object instead of the strings.